### ISSUE TITLE
Strip color/color_temp properties for inactive modes

### ISSUE THREAD

**Issue by sjorge** [Koenkk/zigbee-herdsman-converters#2388]
Ideally we strip the color.<attrib> and color_temp depending on the color_mode, this doesn't seem possible currently?
Returning a new color with with missing attributes just merged this with the old ones.

@Koenkk anyway to do this? Or do we need to add some extra code to `lib/extensions/publish.js` for this?

Also, perhaps we should read colorMode during light configuration to populate it initially?

**Comment by Koenkk** [comment]
I think the first question is if this should be done (note that e.g. the x/y and color_temp are synced with each other).

**Comment by sjorge** [comment]
Not sure, does ha expect this?
If we don’t, we should maybe also sync hs -> color_temp?

~ sjorge

**Comment by Koenkk** [comment]
HA doesn't care about it (https://github.com/home-assistant/core/pull/47993#issuecomment-800862882)

**Comment by sjorge** [comment]
Good, that makes it easier.
Should we still do the hs -> color_temp mapping?

~ sjorge

**Comment by Koenkk** [comment]
> Should we still do the hs -> color_temp mapping?

Yes, I think the logic should be the following: if the current state contains any of the other color modes it should be synced.

e.g. if `meta.state` has `color: {h: XX, s: XX}` and color temp is set, h/s should be updated with the color temp but x/y shouldn't be added.

**Comment by sjorge** [comment]
I've been looking at this a bit thinking about how best to tackle this, some observations.

Small improvements:
- everything needed for conversion seems to be in lib/utils.js, we can create a few wrappers for convenience though
- possible move some of those to lib/light.js as that is a better spot

Issues/questions
- doing the conversion on /set color, color_temp seems relatively easy (perhaps a shared function in lib/light.js that populate this)
- doing the conversion for /set saturation_move, hue_move, color_temp_move are trouble some.
  - we read colorTemperature, currentHue, currentSaturation as needed so we don't have a value to convert from
  - conversion can be done in fromZigbee too... but this has a few issues
    - bulbs that update other values themselves via report --> fight as to what is uses e.g. bulb report hue/sat + color_temp when configure -> innr seems to do a conversion and publish all of them, tint only the active ones BUT will republish the others once maxReportingInterval is reached with old values
    - I would prefer to keep the bulbs reported values as they should be closest to what the bulb is actually set too vs what we calculate from pre-made tables

I guess a complex solution would be to do both:
- pre-convert on set and publish
- set a 'dirty' flag on set of *_move variantes
- when dirty flag set read/reported attribute matches current color_mode, convert and clear dirty flag

That would still allow bulbs to mostly self report for the ones that do, won't cause a bunch of conversions each time.
But we can still do the conversion for *_move once the read returns. This is rather complex though, so I think it's important to agree on a way to tackle this before we work towards this.

**Comment by Koenkk** [comment]
Based on the `color_mode` we can determine what value not to change right? Maybe some bulbs also do the conversion themselves but the calculation is never 1:1 anyway so I don't see a problem also doing the conversion in z2m (fromZigbee)

**Comment by sjorge** [comment]
For the Innr hs -> ct seems pretty accurate.
I’d personally prefer to use what the bulb sends us when possible.

How about having a device_options (default=true) that indicates if we should do the conversion?

~ sjorge

**Comment by Koenkk** [comment]
But when does the device report all these values? (I guess you have reporting enabled?)

**Comment by sjorge** [comment]
Yes (As I have a tint remote and some node-red stuff that does adapting lighthing that uses some of that info)

**Comment by sjorge** [comment]
That's also how i noticed the scene_recall code is not correct atm, as the bulb reports xy and correct x and y values when setting color_temp via a scene (because we do the conversion as nothing seems to support the newer scene color_temp stuff)... even non XY capable bulbs like tradfri WS do. And the reported XY values do match.

**Comment by Koenkk** [comment]
Then I agree a device option should be added. Something like `color_sync`?

**Comment by sjorge** [comment]
> Then I agree a device option should be added. Something like `color_sync`?

Sounds good to me, I'll take a jab at this somewhere during the week, pending how busy I am at work, which impacts how much I want to code in the evenings.

**Comment by sjorge** [comment]
Probably going to be somewhere next week I think, work is pretty hectic at the moment.

**Comment by Koenkk** [comment]
@sjorge no problem, no need to rush things 😄

**Comment by sjorge** [comment]
Been thinking about this in between meetings and redesigning some stuff at work that is taking my time.

```
color_sync: true => convert convert active color(_temp) all modes present in the color state + color_temp
color_sync: legacy => current behavior were we update the set color and color_mode
color_sync: false => we don't update anything (assuming we get it via reporting)
```

Maybe to simplify the `true` case we should also just read color_mode and currentX+currentY or currentHue+currentSaturation depending on what command we used.

We're doing this for the _move variants as we don't have the final color/color_temp for those cases. If we always do a read, we always have the correct value. I did some testing and even hue/innr/tradfri seem to work. Not sure on this yet though but otherwise we'll need convert code in the fromZigbee to handle the _move variants and in toZigbee to handle a plain set. And less duplicate code is better. Although I guess we could just move that to lib/light.js.

 (Upside, uniform with _move variants + simpler tZigbee + accurately reflect what the bulb is displaying. Downside, a read call with 3 attributes)

Maybe in the PR or in a follow up we should look into generating r,g,b too as I think we add those to te payload if we set it using those values. To investigate further if we actually though. the tz.light_color is one of the biggest converters (hence the above thing I've been pondering).

More thinking to do...

**Comment by sjorge** [comment]
I guess we could do both the read and the lib/light.js stuff and end up with

```
color_sync: full => convert convert active color(_temp) all modes present in the color state + color_temp
color_sync: legacy => current behavior were we update the set color and color_mode
color_sync: read  => do a read call to currentX+currentY or currentHue+currentSaturation depending on command we used to set (and still set color_mode)
color_sync: none => we don't update anything (assuming we get it via reporting)
```

**Comment by Koenkk** [comment]
@sjorge I like the color_sync boolean option here, with `true` being the default (this is how it currently is expect that we don't convert all color modes).

**Comment by sjorge** [comment]
> @sjorge I like the color_sync boolean option here, with `true` being the default (this is how it currently is expect that we don't convert all color modes).

So this one:

> 
> ```
> color_sync: true => convert set color(_temp) to all modes present in the color state + color_temp
> color_sync: false => we don't update anything (assuming we get it all via reporting)
> ```

**Comment by Koenkk** [comment]
yes!

**Comment by sjorge** [comment]
@Koenkk how do I get the device options from inside a toZigbee and fromZigbee converter?

**Comment by Koenkk** [comment]
Examples how to get device options:
- fromZigbee: https://github.com/Koenkk/zigbee-herdsman-converters/blob/cfd1fb3fee11d89f25315425249c152a00e60414/converters/fromZigbee.js#L836
- toZigbee: https://github.com/Koenkk/zigbee-herdsman-converters/blob/cfd1fb3fee11d89f25315425249c152a00e60414/converters/toZigbee.js#L222

**Comment by Koenkk** [comment]
@sjorge for both: `meta.state`.

**Comment by sjorge** [comment]
Oops this totally slipped from my radar, I'll see if I can pick this up again soon.
